core/src/main/java/io/grpc/util/GracefulLbSwitch.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.READY;
+
+import io.grpc.ConnectivityState;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A forwarding load balancer and holder of currentLb and pendingLb. The pendingLb's helper will not
+ * update balancing state until a subchannel managed by the pendingLB is READY, whence the pendingLb
+ * becomes to current.
+ */
+@Internal
+@NotThreadSafe // Must be accessed in SynchronizationContext
+public final class GracefulLbSwitch extends ForwardingLoadBalancer {
+
+  // never null
+  private final LoadBalancer newLb;
+  // never null
+  private LoadBalancer currentLb;
+  @Nullable // set to null once the new lb becomes current
+  private LoadBalancer pendingLb;
+
+  /**
+   * Constructor. Example usage:
+   *
+   * <pre>
+   *   private GracefulLbSwitch lbSwitch;
+   *
+   *   void onLbPolicyChange() {
+   *     lbSwitch = new GracefulLbSwitch(lbSwitch, newLbProvider, helper);
```

Comment:
Both the user and this class are doing swap-outs. It might be easier to use if you let the user keep the reference and only this class do the swap-out:
```java
class GracefulSwitchLoadBalancer extends ForwardingLoadBalancer {
  // Assuming changing policy doesn't swap out the child helper
  // If it does, the new helper would be passed here.
  void switchTo(LoadBalancer newLoadBalancer);
  ...
}
```

Reply:
`void switchTo(LoadBalancer newLoadBalancer)` returns void, then there is only one instance of `GracefulSwitchLoadBalancer`? When to shutdown which?

Reply:
+1 to reusing the same instance to doing future switches.

@dapengzhang0, yes, each code location that wants graceful switching would have one instance of the class. I don't understand the shutdown question.

Reply:
@ejona86 `GracefulSwitchLoadBalancer` in @zhangkun83 's code snippet is a `LoadBalancer`, when to shutdown `GracefulSwitchLoadBalancer` instance?

Reply:
When shutdown is called.

Reply:
Then user must not reuse or call `switchTo` after it's shutdown.

Reply:
That seems fine. I'd expect that to largely happen when the LB is being shutdown already.

Reply:
`instance.shutdown()` -> `instance.switchTo(newDelegate)` -> `instance.handleResolvedAddresses()` on the same instance breaks the contract in javadoc of `LoadBalancer#shutdown()`

```
  /**
   * ...  No more methods on this class will be called after this method.  ...
   */
  public abstract void shutdown();
```

Reply:
When you shut down, you throw away the instance. But you're only going to need to shut down the instance when you yourself are shutting down. So there won't be any additional method calls after the shutdown.

Reply:
Done.